tests/e2e/slashing_test.go

```diff
@@ -0,0 +1,307 @@
+
+	// Consumer chain
+	// ====================
+	//
+	// then delegated amount is not updated before the epoch
+	consumerCli.assertTotalDelegated(math.ZeroInt()) // ensure nothing cross staked yet
+
+	// when an epoch ends, the delegation rebalance is triggered
+	consumerCli.ExecNewEpoch()
+
+	// then the total delegated amount is updated
+	consumerCli.assertTotalDelegated(math.NewInt(67_500_000)) // 150_000_000 / 2 * (1 - 0.1)
+
+	// and the delegated amount is updated for the validators
+	consumerCli.assertShare(myExtValidator1, math.LegacyMustNewDecFromStr("45"))   // 100_000_000 / 2 * (1 - 0.1) / 1_000_000 # default sdk factor
+	consumerCli.assertShare(myExtValidator2, math.LegacyMustNewDecFromStr("22.5")) // 50_000_000 / 2 * (1 - 0.1) / 1_000_000 # default sdk factor
+
+	ctx := x.ConsumerChain.GetContext()
+	validator1, found := x.ConsumerApp.StakingKeeper.GetValidator(ctx, myExtValidator1)
+	require.True(t, found)
+	require.False(t, validator1.IsJailed())
+	// FIXME? Off by 1_000_000
+	require.Equal(t, validator1.GetTokens(), sdk.NewInt(46_000_000))
+	validator2, found := x.ConsumerApp.StakingKeeper.GetValidator(ctx, myExtValidator2)
+	require.True(t, found)
+	require.False(t, validator2.IsJailed())
+	// FIXME? Off by 1_000_000
+	require.Equal(t, validator2.GetTokens(), sdk.NewInt(23_500_000))
+
+	// Validator 1 on the Consumer chain is jailed
+	myExtValidator1ConsAddr := sdk.ConsAddress(x.ConsumerChain.Vals.Validators[1].PubKey.Address())
+	jailValidator(t, myExtValidator1ConsAddr, x.Coordinator, x.ConsumerChain, x.ConsumerApp)
+
+	x.ConsumerChain.NextBlock()
+
+	// Assert that the validator's stake has been slashed
+	// and that the validator has been jailed
+	validator1, found = x.ConsumerApp.StakingKeeper.GetValidator(ctx, myExtValidator1)
+	require.True(t, validator1.IsJailed())
+	require.Equal(t, validator1.GetTokens(), sdk.NewInt(36_000_000)) // 10% slash
+
+	// Relay IBC packets to the Provider chain
+	require.NoError(t, x.Coordinator.RelayAndAckPendingPackets(x.IbcPath))
+
+	// Next block on the Provider chain
+	x.ProviderChain.NextBlock()
+
+	// Check new collateral
+	require.Equal(t, 190_000_000, providerCli.QueryVaultBalance())
+	// Check new max lien
+	require.Equal(t, 190_000_000, providerCli.QueryMaxLien())
+	// Check new slashable amount
+	require.Equal(t, 33_000_000, providerCli.QuerySlashableAmount())
+	// Check new free collateral
+	require.Equal(t, 0, providerCli.QueryVaultFreeBalance()) // 190 - max(33, 190) = 190 - 190 = 0
+
+	// Then delegated amount is not updated before the epoch
+	consumerCli.assertTotalDelegated(math.NewInt(67_500_000))
+
+	// When an epoch ends, the slashing accounting on the virtual-staking is triggered
+	consumerCli.ExecNewEpoch()
+	// Relay IBC packets to the Provider chain
+	require.NoError(t, x.Coordinator.RelayAndAckPendingPackets(x.IbcPath))
+
+	// then the total delegated amount is updated
+	// FIXME? No, it's not
```

Comment:
I had stumbled upon this, too. But slashing does not reduce the amount. The name is misleading. It is the total amount minted by the contract.
It is also the amount that was delegated but of course slashing reduces the effectively delegated amount

It is possible to query the current value of the share though:

```go
func (p *TestConsumerClient) assertShareAmount(val sdk.ValAddress, expAmount math.LegacyDec) {
	ctx := p.chain.GetContext()
	del, found := p.app.StakingKeeper.GetDelegation(ctx, p.contracts.staking, val)
	require.True(p.t, found)
	validator, found := p.app.StakingKeeper.GetValidator(ctx, val)
	require.True(p.t, found)
	assert.Equal(p.t, expAmount, validator.TokensFromShares(del.Shares))
}
```

Reply:
👍🏼 Will update the tests, and remove these FIXMEs.